analysers/analyser_osmosis_highway_deadend.py

```diff
@@ -71,121 +71,231 @@
+  NOT is_construction
+"""
+
+sql31 = """
+CREATE INDEX idx_oneway_highway_linestring ON oneway_highway USING GIST(linestring);
+"""
+
+sql32 = """
+-- Collect all nodes that could be possible (valid) end nodes
+CREATE TEMP TABLE allowed_end_nodes AS
+SELECT DISTINCT
+  unnest(nodes) AS id
+FROM
+  ways
+WHERE
+  ways.tags != '' AND
+  (
+    (ways.tags?'amenity' AND ways.tags->'amenity' = 'parking') OR
+    (ways.tags?'railway' AND ways.tags->'railway' = 'platform') OR
+    (ways.tags?'aeroway' AND ways.tags->'aeroway' = 'taxiway')
+  )
+UNION ALL
+SELECT
+  id
+FROM
+  nodes
+WHERE
+  tags != '' AND
+  (
+    (tags?'amenity'  AND tags->'amenity' = 'parking_entrance') OR
+    (tags?'entrance' AND tags->'entrance' IN ('garage', 'emergency'))
+  )
+UNION ALL
+SELECT DISTINCT
+  unnest(nodes) AS id
+FROM
+  highways
+WHERE
+  -- Non-oneway highways are valid input nodes for oneways
+  (NOT is_oneway AND NOT is_roundabout) OR
+  -- Force motorways as valid input nodes
+  -- Raceways are commonly isolated from the main network
+  -- Escape (emergency stop) ways are supposed to be not used / dead-ended
+  -- Footway can lead to anything (attraction entrances, ...) so exclude them even if oneway
+  highway IN ('motorway', 'raceway', 'escape', 'footway') OR
+  -- Construction roads are usually temporary, the connections are likely access=no or similar
+  is_construction
+UNION ALL
+-- Include nodes of ways that cross the extract border. See #1949
+-- Better would be to only include the nodes outside of the extract, but that's expensive
+SELECT
+  unnest(borderways.nodes)
+FROM
+  relation_members AS boundary_members
+  JOIN ways AS boundary_ways ON
+    boundary_members.member_id = boundary_ways.id
+  JOIN highways AS borderways ON
+    ST_Intersects(boundary_ways.linestring, borderways.linestring)
+WHERE
+  boundary_members.member_type = 'W' AND
+  boundary_members.relation_id IN {boundary_ids}
+"""
+
+sql33 = """
+CREATE INDEX idx_allowed_nodes_id ON allowed_end_nodes(id);
+"""
+
+sql34 = """
+-- Make table as short as possible: keep only nodes that are actually nodes of oneways
+CREATE TEMP TABLE allowed_oneway_end_nodes AS
+SELECT DISTINCT
+  allowed_end_nodes.id
+FROM
+  allowed_end_nodes
+  JOIN oneway_highway ON
+    -- Note that we cannot literally only get the first and last node, as i.e. in a looped oneway the 'oneway-entrance/exit node' might be in the middle
+    allowed_end_nodes.id = ANY(oneway_highway.nodes)
+"""
+
+sql35 = """
+CREATE INDEX idx_allowed_oneway_nodes_id ON allowed_oneway_end_nodes(id);
+DROP TABLE allowed_end_nodes;
+"""
+
+sql36 = """
+-- All end nodes of oneway streets
+CREATE TEMP TABLE oneway_terminal AS
+SELECT DISTINCT ON (nid)
+  oneway.id AS wid,
+  terminal.id AS nid,
+  terminal.geom AS geom,
+  CASE
+    WHEN oneway_reverse THEN terminal.id = oneway.nodes[1]
+    ELSE terminal.id = oneway.nodes[oneway.length]
+  END AS is_oneway_last_node
+FROM
+  oneway_highway AS oneway
+  JOIN nodes AS terminal ON
```

Comment:
I think you can remove the join on `nodes` table here. As you only need it for `id`.

The geom can be fetch latter only on selected node ids.

Reply:
I think you're correct. A local implementation works, I'll upload it tomorrow if the results on real data also look good (and fast) :)

Reply:
Done!